notes/memory-and-management.md

```diff
@@ -0,0 +1,666 @@
+</div>
+
+or you can use the overloaded function
+
+<div style="text-align: center">
+
+`type* pointerName = (type *) (malloc((size_t) N, sizeof(type)));`
+
+</div>
+
+**Note:** This only declares the variable, you must initialize the variable in a seperate step. Also, in C, `sizeof()` does not return a `size_t`, so you would have to cast it, along with N.
+
+C++ has an equivalent. You can use the `new` keyword in conjunction with a constructor, just like in Java. However, this works on anything in C++, including primative types. It looks like this:
+
+<div style="text-align: center">
+
+`new Type(arg1, arg2, ..., argN);`
+
+</div>
+
+- Parameter: Corresponds to the parameter of the constructor for Type. For primitive and advanced variables, you may leave it blank or use the desired value corresponding to Type.
+- Return: A Type pointer, or Type*.
+- Use: Typically assigned to a pointer that holds that address.
+
+The assignment is the same as in Java, except the recieving variable MUST be a pointer, just like in C
+
+<div style="text-align: center">
+
+`Type* pointerName = new Type(arg1, arg2, ..., argN);`
+
+</div>
+
+The C++ style of doing this has several advantages:
+
+- Strongly Typed: The return of the C++ statement is more type-safe than the C counter part, as it explicitly labels the pointer as a specific type.
+- Automatic Memory Calcuation: You do not need to actually specify how big of a memory size you want.
+- Faster Initialization: The use of the constructor allows for declaration and initialization all in 1 step.
+
+However, the need for any of these stem from the fact that C/C++ lack a complete garbage collector, with C lacking any concept of it at all. A garbage collector is a system that automatically manages a system's memory. In particular, both systems lack a system that resets destroyed variables, which involves the resetting of the corresponding memory to its null, or zero value.
+
+A use of C++ style allocation can avoid this by providing initialization values in the constructor, but this cannot be done in many cases. So, what happens in most cases is that when `malloc()` or `new` is used, the resulting pointer points to memory containing the value of the previous variable that used that memory. 
+
+This presents an issue in many contexts, for example, when you need to make a counter variable. Fortunately, there are many solutions, one of which is to just set the variable to its zero or null value before use. The easiest, however, is to use:
+
+<div style="text-align: center">
+
+ `void* calloc(size_t size);` 
+ 
+ or 
+ 
+ `void* calloc(size_t N, size_t size);`
+
+</div>
+
+This is the same as `malloc()`, except it resets the value of the memory it allocates before returning it. As you can tell, this originates from C and not C++.
+
+Say you also want to change the size of a variable that has previously been allocated. You must then use:
+
+<div style="text-align: center">
+
+`void* realloc(void *ptr, size_t new_size);`
```

Comment:
lol might be worth adding a note that the use of `realloc` is strongly discouraged by just about every documentation out there

Reply:
Mainly the problem with `realloc` is that it can perform way too many tasks (allocate memory, deallocate memory, reallocate memory, move a block of memory entirely (basically combination of allocate+deallocate), etc.)

This means that it's super easy to mess up with `realloc`, generally some other combination of functions is better

Reply:
Its all over LOST though...